### ISSUE TITLE
Document Architecture Decisions

### ISSUE THREAD

**Issue by drewpc** [redfin/react-server#813]
This is also probably a good time to recommend documenting architecture decisions in the project.  The project itself is very opinionated (which is great!), so all of that should be documented to help explain the reasons behind the decisions.  I'd recommend reading these:

- http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions
- https://github.com/joelparkerhenderson/architecture_decision_record

There's a lot of churn in the open issues and PRs about HTTP/HTTPS servers, applications like Express, Koa, Hapi.js, etc, CORS, ReactServerAgent using SuperAgent or fetch, Webpack 1 vs Webpack 2, isomorphic Webpack, embedding Webpack Dev Server, multiple processes....and that's just stuff I've been involved in!

Now that more and more people are using this, decisions to "should we include Webpack Dev Server for client side HMR?" greatly impacts decisions to "can we support Koa instead of Express?"

**Comment by gigabo** [comment]
Yeah, it would be nice to have a document to point to instead of continually re-hashing things.

Good idea @drewpc.
